Fixes #10480 - Email reporting for discovered hosts

shlomizadok opened the pull request.

shlomizadok committed:
Fixes #10480 - Email reporting for discovered hosts

lzap commented on app/mailers/discovered_mailer.rb:
```diff
@@ -0,0 +1,13 @@
+  def discovered_summary(options = {})
+    raise ::Foreman::Exception.new(N_("Must specify a user with email enabled")) unless (user = User.find(options[:user])) && user.mail_enabled?
+    set_url
+    @time = options[:time] || 'yesterday'
```
Why "yesterday", isn't `Time.now` better?

lzap replied:
Well it might work looking at the code below, if that really works :-)

shlomizadok replied:
It does :)

lzap replied:
It shows as "Summary from yesterday ago to now" in the report. Maybe we can do "1 day" here instead.

shlomizadok replied:
kk, will change :)

shlomizadok replied:
Now shows as `Summary from 2015-11-26 16:10:30 UTC ago to now`

lzap replied:
Hmmm can we do better? There is no "now" when you read an old email, also yesterday and today are a bit fuzzy. Can we do: "Discovered hosts between 2015-11-26 16:10:30 UTC and 2015-11-27 16:10:30 UTC"? Or even better, use user's timezone instead.

lzap commented:
Ok the code looks good, two points :-)
- We need a plain text version for sure (I hate HTML emails)
- It does not work :-)

```
[5] pry(#<DiscoveredMailer>)> Host::Discovered.authorized_as(user, :view_discovered_hosts, Host::Discovered).search_for("created_at >= #{@time}")

=> []
[6] pry(#<DiscoveredMailer>)> @time
=> Thu, 26 Nov 2015 16:10:30 UTC +00:00
[7] pry(#<DiscoveredMailer>)> 
2015-11-26T17:14:33 [app] [I]   Rendered /home/lzap/work/foreman_discovery/app/views/discovered_mailer/discovered_summary.html.erb (5.5ms)
2015-11-26T17:14:33 [app] [I] 
 | Sent mail to lukas@zapletalovi.com (10.4ms)
2015-11-26T17:14:33 [app] [D] Date: Thu, 26 Nov 2015 17:14:33 +0100
 | From: lzap-xxx@redhat.com
 | To: lukas-xxx@zapletalovi.com
 | Message-ID: <56572fe9e01ee_22ab2afdc2041e98604b9@lzapx.brq.redhat.com.mail>
 | Subject: [foreman] Discovered hosts summary
 | Mime-Version: 1.0
 | Content-Type: text/html;
 |  charset=UTF-8
 | Content-Transfer-Encoding: 7bit
 | X-Foreman-Server: 192.168.122.1
 | 
 | <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 | <html>
 | <head>
 |   <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
 |   <title>Summary report for discovered hosts from Foreman</title>
 | </head>
 | <body style="font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; color: #3f3f3f; background-color: #f1f1f1; padding: 10px 24px">
 | <h2 style="font-weight: normal; text-transform: uppercase; font-size: 120%;"><b>Foreman</b> Discovered hosts summary</h2>
 | <h2 style="margin: 5px 0px;">Summary from 2015-11-26 16:10:30 UTC ago to now</h2>
 | <h3 style="margin: 0px;">Discovered hosts from Foreman server at http://192.168.122.1:3000</h3>
 | <div style="background: #fff; padding: 10px 20px; border: 1px solid #e1e2e3">
 |   <table style="min-width: 400px; border: 1px solid #cccccc; padding: 5px; background-color: #ffffff; border-collapse: collapse; text-align: justify; width: 100%;">
 |     <th style="text-align: left; border-collapse: collapse;  background: #f5f5f5; background: linear-gradient(#f5f5f5, #e8e8e8); padding: 4px; border: 1px solid #cccccc;">Name</th>
 |     <th style="text-align: left; border-collapse: collapse;  background: #f5f5f5; background: linear-gradient(#f5f5f5, #e8e8e8); padding: 4px; border: 1px solid #cccccc;">Model</th>
 |     <th style="text-align: left; border-collapse: collapse;  background: #f5f5f5; background: linear-gradient(#f5f5f5, #e8e8e8); padding: 4px; border: 1px solid #cccccc;">IP Address</th>
 |     <th style="text-align: left; border-collapse: collapse;  background: #f5f5f5; background: linear-gradient(#f5f5f5, #e8e8e8); padding: 4px; border: 1px solid #cccccc;">CPUs</th>
 |     <th style="text-align: left; border-collapse: collapse;  background: #f5f5f5; background: linear-gradient(#f5f5f5, #e8e8e8); padding: 4px; border: 1px solid #cccccc;">Memory</th>
 |     <th style="text-align: left; border-collapse: collapse;  background: #f5f5f5; background: linear-gradient(#f5f5f5, #e8e8e8); padding: 4px; border: 1px solid #cccccc;">Disk count</th>
 |     <th style="text-align: left; border-collapse: collapse;  background: #f5f5f5; background: linear-gradient(#f5f5f5, #e8e8e8); padding: 4px; border: 1px solid #cccccc;">Disks size</th>
 |   </table>
 | </div>
 | </body>
 | </html>
 | 
[lzap@lzapx foreman]$ date
Thu Nov 26 17:14:34 CET 2015
```

As you can see when I run `rake reports:daily` the time provided is the current time. So the query should search something like "time - 24 hours". I only configured with "Daily" report, weekly should work as well.

lzap commented:
Or maybe teach me how to test this.

shlomizadok commented:
@lzap - It is sending reports of the discovered hosts created between `period` and now.
So if new discovered hosts were created today, on the next day you'll get an email with their details.
1. Add a new discovered host.
2. run `rake reports:daily`
You should get an email with the list of the new discovered hosts created between yesterday and today.
1. Create another discovered host
   3a. In the UI change the report period to weekly
2. Wait a week
3. run `rake reports:weekly`
   You should get an email with the list of the new discovered hosts created between last week and today.
   :smile:

lzap commented:
Additional issue, when the plugin is disabled, we get this error:

```
NameError: uninitialized constant DiscoveredMailer
/home/lzap/work/foreman/app/models/mail_notification.rb:41:in `deliver'
/home/lzap/work/foreman/app/models/user_mail_notification.rb:18:in `deliver'
/home/lzap/work/foreman/lib/tasks/reports.rake:97:in `block in process_notifications'
/home/lzap/work/foreman/lib/tasks/reports.rake:96:in `process_notifications'
/home/lzap/work/foreman/lib/tasks/reports.rake:102:in `block (2 levels) in <top (required)>'
Tasks: TOP => reports:daily
```

This is a genric issue for all plugins which we should solve. Ideally, MailNotification object should not be stored in the database, but registered in engine.rb files instead. Or a workaround - to skip failed mailers.

shlomizadok commented:
Isn't this solved with https://github.com/theforeman/foreman_discovery/pull/232/files#diff-f18fc5c15182f24c5ce6e80aa954eb7bR1 ?

shlomizadok commented:
@lzap - Added text version as well. Let me know if it plays nice for you.
As for the `NameError: uninitialized constant DiscoveredMailer` - afaik, core and plugins add this via seed

lzap commented:
Yes, that's added via seeds and nobody cares what happens when you disable/uninstall a plugin. It breaks whole emailing subsystem, we need to take care of this. I suggest a patch in core to ignore non-existing mailers.

Tests failed.

shlomizadok commented:
@lzap - I have changed the query (from `search_for` to `where`) as it is faster. I have also added support for user time zone and changed the "now" to date.

shlomizadok commented:
Finally fixed the tests :smile_cat:

lzap commented:
Ok will take a look, reported the mailer error as http://projects.theforeman.org/issues/12645 if you want to take a look.

lzap commented:
I am getting

```
ArgumentError: Invalid Timezone: 
/home/lzap/work/foreman_discovery/app/mailers/discovered_mailer.rb:5:in `discovered_summary'
/home/lzap/work/foreman/app/models/mail_notification.rb:41:in `deliver'
/home/lzap/work/foreman/app/models/user_mail_notification.rb:18:in `deliver'
/home/lzap/work/foreman/lib/tasks/reports.rake:97:in `block in process_notifications'
/home/lzap/work/foreman/lib/tasks/reports.rake:96:in `process_notifications'
/home/lzap/work/foreman/lib/tasks/reports.rake:102:in `block (2 levels) in <top (required)>'
Tasks: TOP => reports:daily
```

with a user which has Browser timezone set.
Also can you please rebase.

lzap commented:
Ping, when you have time, there is this error to solve.

shlomizadok commented:
@lzap - Added a fallback to UTC, let me know if that's okay :)

lzap commented on app/views/discovered_mailer/discovered_summary.text.erb:
```diff
@@ -0,0 +1,24 @@
+<%= _("Foreman discovered hosts summary") %>
+=============================================
```
When this is rendered, it is longer than the text, please remove the tail ones.

```
Foreman discovered hosts summary
=============================================
```

lzap commented:
Ok please my last remark, and do please rebase on top of `develop` (the current version does not boot on Rails 4). Thanks!

shlomizadok commented:
@lzap - thanks for reviewing. rebased. removed the long tail :)

lzap commented:
Great work @shlomizadok !

lzap commented:
Merged as 90fd569089f8b119ae0f5b283a3c130143134472, thank you!

lzap commented:
Remember to present this on the next demo please.
